Only raise Application.create events for Superkey sources when complete

lindgrenj6 opened the pull request:
During the integration process for superkey, I'm finding that ordering of messages is the biggest problem faced.

So I'm reworking to only raise the `Application.create` event once the superkey worker posts back. It's complicated logic, but it's necessary to make sure that the events go out in the right order:

1. `Source.create`
2. `Application.create`
3. `Authentication.create`
4. `ApplicationAuthentication.create`

\# TODO:
- [x] tests

lindgrenj6 commented:
cc @syncrou

syncrou commented on app/controllers/api/v1/applications_controller.rb:
```diff
@@ -4,15 +4,35 @@ class ApplicationsController < ApplicationController
-        raise_event("#{model}.create", application.as_json)
+        # we do not want to raise the create event since the application has
+        # not been processed by the superkey worker.
+        raise_event("Application.create", application.as_json) unless application.source.super_key?
```
I'm totally outside this codebase, but can any of the `Event` concern methods be used here? `app/models/concerns/event_concern.rb` so we could hide some of the conditional logic within a `raise_event` method?

lindgrenj6 replied:
I really like the look of #344, so I'll wait on that one and refactor this to use the `raise_event_if` pattern.

syncrou reviewed (commented):
Behavior looks good. My only nit, can we hide the conditionals for `raise_event` in the `event_concern` at all like what we're doing here: https://github.com/RedHatInsights/sources-api/pull/344/files

I realize I've been out of this code for a while so feel free to let me know if this is not a workable idea.

lindgrenj6 commented:
So it appears some of the logic is backwards for some conditions, e.g. I added a `raise_event_unless` method that inverses the check, but it's still hard to read inverse checks.

examples:
1.
```
raise_event_if(!application.source.super_key?, "Application.create", ...)
```
This reads fine, we're raising if the application is not a super key.


2.
```
first_time_superkey = (stuff resolving to true, indicating that this is the superkey application update)
raise_event_unless(first_time_superkey, "Application.create", application.as_json)
```
This one seems...off, `raise_event_if(first_time_superkey,..)` doesn't do the right behavior, so I added `raise_event_unless`, but that just reads backwards.


3.
```
raise_event_unless(!app.source.super_key?, "Application.create", app.as_json)
```
This one reads completely backwards. But `raise_event_if(!app.source.super_key?, ...)` doesn't work, and `raise_event_unless(app.source.super_key?, ...)` doesn't work either. 
So we have the double negative and it works. 

I'm not sure why it is this way, I'll keep playing with it some more.

slemrmartin commented:
What about this form:
```
not_superkey = !app.source.superkey?
raise_event_if(not_superkey, ...)
```

lindgrenj6 commented:
@slemrmartin I added a commit that just reverses the `raise_event_if` logic.

Now it reads like this:
`raise_event_if(truthy_val, message, values, headers)`

This makes the events read a little better imo since if is usually checking for something that resolves to `true`. 

Let me know what you think. cc @lpichler

lpichler commented on app/models/concerns/event_concern.rb:
```diff
@@ -8,16 +8,16 @@ module EventConcern
   def ignore_raise_event_for?(attributes)
     ignore_attribute_list = IGNORE_RAISE_EVENT_LIST[self.class.name]
-    return false unless ignore_attribute_list
+    return true unless ignore_attribute_list
```
@lindgrenj6 Should not we neg also name of this method in this case?
something like `ignore_raise_event_for` -> `allow_raise_event_or`.
Original intention was - what is in `IGNORE_RAISE_EVENT_LIST` it should be ignored
so If model is not present  in `IGNORE_RAISE_EVENT_LIST` we should not ignore raising event and it means we should raise event.

lindgrenj6 replied:
Absolutely, I renamed the method to account for the reversed behavior. Sorry I missed that.

lpichler commented on app/models/concerns/event_concern.rb:
```diff
@@ -8,16 +8,16 @@ module EventConcern
+    return true unless ignore_attribute_list
 
-    (IGNORE_RAISE_EVENT_ATTRIBUTES_LIST & attributes.map(&:to_sym)).present?
+    (ignore_attribute_list & attributes.map(&:to_sym)).empty?
```
thanks for `IGNORE_RAISE_EVENT_LIST` -> `ignore_attribute_list` 👍

lpichler commented on app/controllers/application_controller.rb:
```diff
@@ -74,15 +74,15 @@ def instance_link(instance)
-  def raise_event_if(ignore_raise_event, event, payload)
-    return if ignore_raise_event
+  def raise_event_if(raise_event, event, payload)
+    return unless raise_event
```
yes `ignore_raise_event` was not great but what about to name this more like `raise_event_allowed` or  `raise_event_expression` or `raise_event_condition`
or something like that express that it is condition

lindgrenj6 replied:
Changed to `raise_event_allowed` :)

lpichler commented:
> @slemrmartin I added a commit that just reverses the `raise_event_if` logic.
> 
> Now it reads like this:
> `raise_event_if(truthy_val, message, values, headers)`
> 
> This makes the events read a little better imo since if is usually checking for something that resolves to `true`.
> 
> Let me know what you think. cc @lpichler

I added some suggestions but they make sense to you but yes naming is hard. Otherwise looks good 👍 
thanks for reducing "if"s.

syncrou reviewed (approved):
LGTM 👍 - with the current tacit approval from @lpichler @slemrmartin I am merging this so we can get input from the app teams.

If either of you feel this should have additional changes please raise with @lindgrenj6 and we'll get them in followup PRs.
